### PR TITLE
Glue: Show data links only for fully interpolated correlations

### LINKED ISSUE #58574: Glue: Change interpolation logic to return a link only if all the variables can be interpolated
Follow up on #58235

The link should be created only if all variables were interpolated in the target query.

For example:

* source data source: Loki
* target data source: Prometheus
* target query: `{ app="${appName}", host="${hostName}" }`
* link attached to field: `appName`

(1) Given the following data frames returned by Loki:

| time | appName | hostName | msg |
|----|----|-----|----|
| 1 | foo | host001 | deployed |

The link should be shown to `appName` field.

(2) Given the following data frames returned by Loki:

| time | appName | hostName |msg |
|----|----|----|---|
| 1 | foo |  | app error |

The link should *not* be shown because the `hostName` field value is missing for the interpolation.

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

The problem is described in https://github.com/grafana/grafana/issues/58574.

**Which issue(s) this PR fixes**:

Fixes #58574

**Special notes for your reviewer**:

The logic takes care of returning a link model only if the query can be fully interpolated. 

The additional filtering is implemented as "data links filters" which could be refactored in the future, so they are added in runtime. For now, they handle permissions and interpolation.

<details><summary>How to test it</summary>

Ensure the feature toggle for correlations is on in custom.ini:

```ini
[feature_toggles]
correlations = true
```

1. Install and run mock data source

2. Create a new correlation from Mock to any data source using variables for `app` and `endpoint` fields, and attach it to `app` field e.g.:  Prometheus `{app="${__data.fields.app}", endpoint="${__data.fields.endpoint}"}`:
Screenshot 2022-11-17 at 13 47 31

3. Run mock data source using raw data frame and paste [the following data frame](https://gist.githubusercontent.com/ifrost/8c1d5df13681aa8bce1216bd7cfa0fb8/raw/796183ef58e5517e2e7147cab64c5766d4f4a97c/logs.json)

4. Run the query

5. Expected result: the link to the correlations should show up only in the log line with both fields provided:
Screenshot 2022-11-17 at 13 34 48

6. Expected result: clicking on the link should open target data source with interpolated values

</details>

[^1]: alternatively, to support our use case, we could create a separate instance of TemplateService inside `getFieldLinksForExplore` and initialize it with scoped variables.

### CODE REVIEW

File: public/app/features/explore/utils/links.ts [hunk 1]
```diff
@@ -10,12 +10,42 @@ import {
...
+};
+
+const dataLinkHasAllVariablesDefined = (link: DataLink, scopedVars: ScopedVars) => {
+  let hasAllRequiredVarsAvailable = true;
+
+  if (link.internal) {
+    let stringifiedQuery = '';
+    try {
+      stringifiedQuery = JSON.stringify(link.internal.query || {});
+      getTemplateSrv().replace(stringifiedQuery, scopedVars, (f: string) => {
```
Comment (gelicia) [inline public/app/features/explore/utils/links.ts:33]: Can we get a comment on what the format function is doing here?

File: public/app/features/explore/utils/links.ts [hunk 2]
```diff
@@ -56,13 +86,9 @@ export const getFieldLinksForExplore = (options: {
...
   if (field.config.links) {
-    const links = [];
-
-    if (!contextSrv.hasAccessToExplore()) {
-      links.push(...field.config.links.filter((l) => !l.internal));
-    } else {
-      links.push(...field.config.links);
-    }
+    const links = field.config.links.filter((link) => {
+      return DATA_LINK_FILTERS.every((filter) => filter(link, scopedVars));
```
Comment (gelicia) [inline public/app/features/explore/utils/links.ts:90]: clever!

File: public/app/features/explore/utils/links.test.ts [hunk 3]
```diff
@@ -124,8 +122,39 @@ describe('getFieldLinksForExplore', () => {
     const links = getFieldLinksForExplore({ field, rowIndex: 0, range });
     expect(links).toHaveLength(0);
   });
+
+  it('returns internal links when target contains defined template variables', () => {
```
Comment (gelicia) [inline public/app/features/explore/utils/links.test.ts:126]: Can a comment explain what rowIndex points to here, such that the setup for these are the same but rowIndex 1 is a test for one condition and 0 is another?

**Review by gelicia — APPROVED** [review]
This works well, I followed your provided testing steps. Some of the logic is a little obscure, some comments would help but are not necessary to merge. Good work!

File: public/app/features/explore/utils/links.test.ts (hunk 3 above)
Comment (ifrost) [reply public/app/features/explore/utils/links.test.ts:136]: Good point. I've changed the code to use more explicit constant names in 1d997cf80fc2aa02a67e8e1efe87e53d160f141c. I hope it's more clear now.

**Pull request merged**